### PR TITLE
[Feature] Support panoptic segmentation result analysis

### PR DESCRIPTION
**Pull request by chhluo**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

1. Support panoptic segmentation result analysis.  The total evaluation time on results (coco val2017) generated by mask2former_r50 takes around 6 mins.

bad sample
000000017905.jpg with PQ = 0.308

good sample
000000455624.jpg with PQ = 0.887

2. add args `overlay_gt_pred`. image with gts and image with predictions will be concated.

set `overlay_gt_pred = False`, to plot gts and predictions on the images separately.

set `overlay_gt_pred = True`, to plot gts and predictions on the same images.

000000104612_0 207_overlay

## Modification

1. Add `print_log` in https://github.com/chhluo/mmdetection/blob/89620985c9c535854011283900d994b2925f1904/mmdet/datasets/api_wrappers/panoptic_evaluation.py#L29, to control whether to print log. In analyze result, log is unnecessary when we only evaluate one image a time.
2. Add [panoptic_evaluate](https://github.com/chhluo/mmdetection/blob/89620985c9c535854011283900d994b2925f1904/tools/analysis_tools/analyze_results.py#L209) and refactor `ResultVisualizer`.

3. add `overlay_gt_pred` in imshow_gt_det_bboxes.
 
## BC-breaking (Optional)

Does the modification introduce changes that break the backward-compatibility of the downstream repos?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
3. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
5. If the modification has potential influence on downstream projects, this PR should be tested with downstream projects, like MMDet or MMCls.
6. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

**Comment by ZwwWayne** [comment]
considering draw GT and prediction in two images for clear comparison

File: mmdet/core/visualization/image.py [hunk 1]
```diff
@@ -425,6 +426,8 @@ def imshow_gt_det_bboxes(img,
       wait_time (float): Value of waitKey param. Default: 0.
       out_file (str, optional): The filename to write the image.
           Default: None.
+      overlay_gt_pred (bool): Whether to plot gts and predictions on the
```
Comment (hhaAndroid) [inline mmdet/core/visualization/image.py:429]: It needs to be explained how to arrange if it is false.

File: tools/analysis_tools/analyze_results.py [hunk 2]
```diff
@@ -68,22 +69,44 @@ class ResultVisualizer:
     """Display and save evaluation results.
 
     Args:
-        show (bool): Whether to show the image. Default: True
+        show (bool): Whether to show the image. Default: True.
         wait_time (float): Value of waitKey param. Default: 0.
         score_thr (float): Minimum score of bboxes to be shown.
-           Default: 0
+           Default: 0.
+        overlay_gt_pred (bool): Whether to plot gts and
```
Comment (hhaAndroid) [inline tools/analysis_tools/analyze_results.py:76]: same as above

File: tools/analysis_tools/analyze_results.py [hunk 3]
```diff
@@ -183,6 +300,11 @@ def parse_args():
         type=float,
         default=0,
         help='score threshold (default: 0.)')
+    parser.add_argument(
+        '--overlay-gt-pred',
+        action='store_true',
+        help='whether to plot gts and predictions'
```
Comment (hhaAndroid) [inline tools/analysis_tools/analyze_results.py:306]: same as above

File: tools/analysis_tools/analyze_results.py [hunk 4]
```diff
@@ -220,7 +342,8 @@ def main():
     outputs = mmcv.load(args.prediction_path)
 
     result_visualizer = ResultVisualizer(args.show, args.wait_time,
```
Comment (hhaAndroid) [inline tools/analysis_tools/analyze_results.py:344]: Please also verify that the effect of `demo/create_result_gif.py` is correct when `overlay-gt-pred=False`

File: tools/analysis_tools/analyze_results.py [hunk 5]
```diff
@@ -68,22 +69,44 @@ class ResultVisualizer:
...
+                                dataset,
+                                results,
+                                performances,
+                                out_dir=None):
+        """Display or save image with groung truths and predictions from a
+        model.
+
+        Args:
+            dataset (Dataset): A PyTorch dataset.
+            results (list): Results from test results pkl file.
```
Comment (hhaAndroid) [inline tools/analysis_tools/analyze_results.py:100]: Object detection or panoptic segmentation results from test results pkl file.

File: tools/analysis_tools/analyze_results.py [hunk 6]
```diff
@@ -68,22 +69,44 @@ class ResultVisualizer:
...
+            performances (dict): A dict contains samples's indices
+                in dataset and model's performance on them.
+            out_dir (str, optional): The filename to write the image.
+                Defaults to None.
```
Comment (hhaAndroid) [inline tools/analysis_tools/analyze_results.py:104]: Default: None

File: mmdet/datasets/api_wrappers/panoptic_evaluation.py [hunk 7]
```diff
@@ -25,7 +25,8 @@ def pq_compute_single_core(proc_id,
                            gt_folder,
                            pred_folder,
                            categories,
-                           file_client=None):
+                           file_client=None,
+                           print_log=True):
```
Comment (ZwwWayne) [inline mmdet/datasets/api_wrappers/panoptic_evaluation.py:29]: suggest use False by default

File: tools/analysis_tools/analyze_results.py (hunk 4 above)
Comment (chhluo) [reply tools/analysis_tools/analyze_results.py:351]: I will do it in the next pr.

**Comment by hhaAndroid** [comment]
@chhluo Can you fix this [bug](https://github.com/open-mmlab/mmdetection/issues/8035) at the same time

**Review by hhaAndroid — APPROVED** [review]

File: mmdet/core/visualization/image.py [hunk 8]
```diff
@@ -505,20 +511,49 @@ def imshow_gt_det_bboxes(img,
...
+            bbox_color=det_bbox_color,
+            text_color=det_text_color,
+            mask_color=det_mask_color,
+            thickness=thickness,
+            font_size=font_size,
+            win_name=win_name,
+            show=False)
+        img = np.concatenate([img_with_gt, img_with_det], axis=0)
+
+        plt.imshow(img)
```
Comment (ZwwWayne) [inline mmdet/core/visualization/image.py:548]: why the show logic is only allowed when overlay_gt_pred=False? show should work with both cases.

Comment (chhluo) [reply mmdet/core/visualization/image.py:548]: when overlay_gt_pred=True, the show logic is in `imshow_det_bboxes`, https://github.com/open-mmlab/mmdetection/blob/240d7a31c745578aa8c4df54c3074ce78b690c34/mmdet/core/visualization/image.py#L358

**Pull request merged**